### PR TITLE
ADM-999 [frontend][backend][docs]: add api for source control configuration in the metrics page

### PR DESCRIPTION
**Pull request by zhou-yinyuan**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/controller/source/SourceController.java [hunk 1]
```diff
@@ -53,16 +58,55 @@ public ResponseEntity<Void> verifyBranch(
...
-			}
-			default -> {
-			}
-		}
+		gitHubService.verifyCanReadTargetBranch(request.getRepository(), request.getBranch(), request.getToken());
+		log.info("Successfully verify source type: {} branch: {}.", sourceType, request.getBranch());
 		return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
 	}
 
+	@PostMapping("/{sourceType}/organizations")
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/source/SourceController.java:66]: 这是contrller， 我们可不可复用buildkit 的接口？
if no, why?

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 2]
```diff
@@ -333,4 +363,174 @@ public CommitInfo fetchCommitInfo(String commitId, String repositoryId, String t
...
 	}
 
+	public List<String> getAllOrganizations(String token) {
+		int initPage = 1;
+		String realToken = BEARER_TITLE + token;
+		PageOrganizationsInfoDTO pageOrganizationsInfoDTO = cachePageService.getGitHubOrganizations(realToken, initPage,
+				PER_PAGE);
+		List<OrganizationsInfoDTO> firstPageStepsInfo = pageOrganizationsInfoDTO.getPageInfo();
+		int totalPage = pageOrganizationsInfoDTO.getTotalPage();
+		log.info("Successfully parse the total page_total page of organizations: {}", totalPage);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:373]: no log of `start`?

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 3]
```diff
@@ -333,4 +363,174 @@ public CommitInfo fetchCommitInfo(String commitId, String repositoryId, String t
...
+	}
+
+	public List<String> getAllRepos(String token, String organization) {
+		int initPage = 1;
+		String realToken = BEARER_TITLE + token;
+		PageReposInfoDTO pageReposInfoDTO = cachePageService.getGitHubRepos(realToken, organization, initPage,
+				PER_PAGE);
+		List<ReposInfoDTO> firstPageStepsInfo = pageReposInfoDTO.getPageInfo();
+		int totalPage = pageReposInfoDTO.getTotalPage();
+		log.info("Successfully parse the total page_total page of repos: {}", totalPage);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:405]: no log of start

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 4]
```diff
@@ -333,4 +363,174 @@ public CommitInfo fetchCommitInfo(String commitId, String repositoryId, String t
...
+	}
+
+	public List<String> getAllBranches(String token, String organization, String repo) {
+		int initPage = 1;
+		String realToken = BEARER_TITLE + token;
+		PageBranchesInfoDTO pageBranchesInfoDTO = cachePageService.getGitHubBranches(realToken, organization, repo,
+				initPage, PER_PAGE);
+		List<BranchesInfoDTO> firstPageStepsInfo = pageBranchesInfoDTO.getPageInfo();
+		int totalPage = pageBranchesInfoDTO.getTotalPage();
+		log.info("Successfully parse the total page_total page of branches: {}", totalPage);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:438]: start log

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 5]
```diff
@@ -333,4 +363,174 @@ public CommitInfo fetchCommitInfo(String commitId, String repositoryId, String t
...
+
+	public List<String> getAllCrews(String token, String organization, String repo, String branch, long startTime,
+			long endTime) {
+		int initPage = 1;
+		String realToken = BEARER_TITLE + token;
+		PagePullRequestInfoDTO pageBranchesInfoDTO = cachePageService.getGitHubPullRequest(realToken, organization,
+				repo, branch, initPage, PER_PAGE);
+		List<PullRequestInfoDTO> firstPageStepsInfo = pageBranchesInfoDTO.getPageInfo();
+		int totalPage = pageBranchesInfoDTO.getTotalPage();
+		log.info("Successfully parse the total page_total page of pull requests: {}", totalPage);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:473]: start log

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/CachePageServiceTest.java [hunk 6]
```diff
@@ -184,13 +199,76 @@ void shouldReturnPagePipelineInfoDtoWhenFetchPagePipelineInfoSuccessGivenExistBu
...
+		assertThat(pageBranchesInfoDTO.getPageInfo()).isEqualTo(responseEntity.getBody());
+		assertThat(pageBranchesInfoDTO.getTotalPage()).isEqualTo(2);
+	}
+
+	@Test
+	void shouldReturnPagePullRequestInfoDtoWhenFetchPullRequestInfoSuccessGivenExist() throws IOException {
+		String organization = "test-org";
+		String repo = "test-repo";
+		String branch = "test-branch";
+		HttpHeaders httpHeaders = buildHttpHeaders(GITHUB_TOTAL_PAGE_HEADER);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/pipeline/buildkite/CachePageServiceTest.java:253]: good testing pattern

but, duplicated code

File: backend/src/main/java/heartbeat/controller/source/SourceController.java (hunk 1 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/controller/source/SourceController.java:66]: 不太好用，因为按照现在的逻辑，buildkite的token都没传，自然不可能使用buildkite的接口去拿数据了

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java (hunk 2 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/service/source/github/GitHubService.java:373]: done

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java (hunk 4 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/service/source/github/GitHubService.java:438]: done

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java (hunk 3 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/service/source/github/GitHubService.java:405]: done

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java (hunk 5 above)
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/service/source/github/GitHubService.java:473]: done

File: backend/src/test/java/heartbeat/service/pipeline/buildkite/CachePageServiceTest.java (hunk 6 above)
Comment (zhou-yinyuan) [reply backend/src/test/java/heartbeat/service/pipeline/buildkite/CachePageServiceTest.java:253]: done

**Review by Yufan0226 — APPROVED** [review]

**Pull request merged**